In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('./data/titanic.csv')
df.shape

(891, 12)

In [57]:
df.isnull().sum()

PassengerId          0
Survived             0
Pclass               0
Name                 0
Sex                  0
Age                177
SibSp                0
Parch                0
Ticket               0
Fare                 0
Cabin              687
Embarked             2
survival_gender      0
dtype: int64

In [15]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [174]:
df.groupby(['Pclass']).PassengerId.count()

Pclass
1    216
2    184
3    491
Name: PassengerId, dtype: int64

In [6]:
df.groupby('Pclass').Fare.mean()

Pclass
1    84.154687
2    20.662183
3    13.675550
Name: Fare, dtype: float64

In [25]:
df.groupby('Sex').Age.mean()

Sex
female    27.915709
male      30.726645
Name: Age, dtype: float64

In [26]:
df.groupby(['Sex', 'Pclass']).Age.mean()

Sex     Pclass
female  1         34.611765
        2         28.722973
        3         21.750000
male    1         41.281386
        2         30.740707
        3         26.507589
Name: Age, dtype: float64

In [24]:
age = df.groupby('Sex').Age.agg([len,min,max]).reset_index()
age

Sex  len   min   max
0  female  314  0.75  63.0
1    male  577  0.42  80.0

#### 1. Who were the passengers?

Most passengers were in Pclass 3(491), Pclass 2 had 184 passengers, and Pclass 1 had 216 passengers.\
Female:\
youngest is almost a year old and the oldest is 63.\
Male:\
youngest is less than half a year old and the oldest is 80.\
The youngest group of people belonged to the third class for male and female.

In [4]:
df['deck']= df['Cabin'].str[:1]
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked deck  
0      0         A/5 21171   7.2500   NaN        S  NaN  
1      0          PC 17599  71.2833   C85        C    C  
2      0  STON/O2. 3101282   7.9250   NaN        S  NaN  
3      0            113803  53.1000  C123        S    C  
4      0            373450   8.0500   NaN        S  NaN

In [6]:
df.deck.value_counts()

C    59
B    47
D    33
E    32
A    15
F    13
G     4
T     1
Name: deck, dtype: int64

In [173]:
df.groupby(['Pclass', 'deck']).PassengerId.count()

Pclass  deck
1       A       15
        B       47
        C       59
        D       29
        E       25
        T        1
2       D        4
        E        4
        F        8
3       E        3
        F        5
        G        4
Name: PassengerId, dtype: int64

2. What deck were the passengers on and how's that relate to their class booking?\
Many passenger's did not have cabin's reported, but what was reported showed that Pclass 1 stayed mainly on Deck C and B followed by D, E, A, and T respectively.\ 
Pclass 2 stayed in F, E, and D; Pclass 3 stayed in F, G, and E.\ 
Deck C and B had the h ighest average fare price and G was the lowest. 

In [31]:
df.groupby(['Pclass','deck']).Fare.mean()

Pclass  deck
1       A        39.623887
        B       113.505764
        C       100.151341
        D        63.324286
        E        55.740168
        T        35.500000
2       D        13.166675
        E        11.587500
        F        23.750000
3       E        11.000000
        F        10.611660
        G        13.581250
Name: Fare, dtype: float64

In [42]:
df.groupby(['Embarked']).count()

PassengerId  Survived  Pclass  Name  Sex  Age  SibSp  Parch  Ticket  \
Embarked                                                                        
C                 168       168     168   168  168  130    168    168     168   
Q                  77        77      77    77   77   28     77     77      77   
S                 644       644     644   644  644  554    644    644     644   

          Fare  Cabin  deck  
Embarked                     
C          168     69    69  
Q           77      4     4  
S          644    129   129

In [43]:
df.groupby(['Pclass', 'Embarked']).count()

PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  \
Pclass Embarked                                                                
1      C                  85        85    85   85   74     85     85      85   
       Q                   2         2     2    2    2      2      2       2   
       S                 127       127   127  127  108    127    127     127   
2      C                  17        17    17   17   15     17     17      17   
       Q                   3         3     3    3    2      3      3       3   
       S                 164       164   164  164  156    164    164     164   
3      C                  66        66    66   66   41     66     66      66   
       Q                  72        72    72   72   24     72     72      72   
       S                 353       353   353  353  290    353    353     353   

                 Fare  Cabin  deck  
Pclass Embarked                     
1      C           85     66    66  
       Q            2      2     2  
       S          127    106   106  
2      C           17      2     2  
       Q            3      1     1  
       S          164     13    13  
3      C           66      1     1  
       Q           72      1     1  
       S          353     10    10

S=Southampton\
C=Cherbourg\
Q=Queenstown

In [77]:
df.Embarked.value_counts(normalize = True)*100

S    72.440945
C    18.897638
Q     8.661417
Name: Embarked, dtype: float64

In [58]:
df.Embarked.value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

In [20]:
df.groupby(['Pclass', 'Embarked']).PassengerId.count()

Pclass  Embarked
1       C            85
        Q             2
        S           127
2       C            17
        Q             3
        S           164
3       C            66
        Q            72
        S           353
Name: PassengerId, dtype: int64

3. Where the passengers come from?
Around 72.4%(644 passengers) of our passengers came from Southampton, 18.9%(168 passengers) came from Cherboug, and 8.8%(77 passengers) came from Queenstown.\
Most people came from Southampton and are Pclass3, Pclass2, and Pclass1 respectively.

In [49]:
df.Sex.describe()

count      891
unique       2
top       male
freq       577
Name: Sex, dtype: object

In [33]:
def data(row):
    if row.Survived == 1 and row.Sex == 'female':
        return 'female survived'
    elif row.Survived == 1 and row.Sex =='male':
        return 'male survived'
    elif row.Survived == 0 and row.Sex == 'female':
        return 'female died'
    else:
        return 'male died'

In [34]:
df['survival_gender'] = df.apply(data, axis='columns')

In [34]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  survival_gender  
0      0         A/5 21171   7.2500   NaN        S        male died  
1      0          PC 17599  71.2833   C85        C  female survived  
2      0  STON/O2. 3101282   7.9250   NaN        S  female survived  
3      0            113803  53.1000  C123        S  female survived  
4      0            373450   8.0500   NaN        S        male died

In [35]:
df.survival_gender.value_counts(normalize = True)

male died          0.525253
female survived    0.261504
male survived      0.122334
female died        0.090909
Name: survival_gender, dtype: float64

In [ ]:
df.survival_gender.describe()

count           891
unique            4
top       male died
freq            468
Name: survival_gender, dtype: object

In [43]:
df.survival_gender.value_counts()

male died          468
female survived    233
male survived      109
female died         81
Name: survival_gender, dtype: int64

In [59]:
df.groupby('survival_gender').Fare.mean()

survival_gender
female died        23.024385
female survived    51.938573
male died          21.960993
male survived      40.821484
Name: Fare, dtype: float64

In [79]:
df.groupby(['survival_gender','Pclass']).PassengerId.count()

survival_gender  Pclass
female died      1           3
                 2           6
                 3          72
female survived  1          91
                 2          70
                 3          72
male died        1          77
                 2          91
                 3         300
male survived    1          45
                 2          17
                 3          47
Name: PassengerId, dtype: int64

In [24]:
df.groupby(['Survived', 'Pclass','Sex']).PassengerId.count()

Survived  Pclass  Sex   
0         1       female      3
                  male       77
          2       female      6
                  male       91
          3       female     72
                  male      300
1         1       female     91
                  male       45
          2       female     70
                  male       17
          3       female     72
                  male       47
Name: PassengerId, dtype: int64

In [16]:
df.groupby(['Survived', 'Pclass']).Age.mean()

Survived  Pclass
0         1         43.695312
          2         33.544444
          3         26.555556
1         1         35.368197
          2         25.901566
          3         20.646118
Name: Age, dtype: float64

1.What factors helped them survive?\
Being a woman and having family members had the largest impact on whether you were going to survive.\
233 women survived:\
Pclass1: 91, Pclass2: 70, Pclass3: 72\
Survived with family: 134 women, 64 men\
Died with family: 121 men, 27 women\
109 males survived(men who survived had younger average age compared to women who survived):\
Pclass1: 45, Pclass2: 17, Pclass3: 47. 64 men traveling alone and 45 men traveling with family survived.

In [111]:
df.head(1)

PassengerId  Survived  Pclass                     Name   Sex   Age  SibSp  \
0            1         0       3  Braund, Mr. Owen Harris  male  22.0      1   

   Parch     Ticket  Fare Cabin Embarked survival_gender deck  
0      0  A/5 21171  7.25   NaN        S       male died  NaN

In [30]:
df.groupby(['Survived', 'Pclass']).deck.value_counts()

Survived  Pclass  deck
0         1       C       24
                  B       12
                  A        8
                  D        7
                  E        7
                  T        1
          2       D        1
                  E        1
                  F        1
          3       F        4
                  G        2
1         1       B       35
                  C       35
                  D       22
                  E       18
                  A        7
          2       F        7
                  D        3
                  E        3
          3       E        3
                  G        2
                  F        1
Name: deck, dtype: int64

There are 687 nulls in the cabin(deck) field which does impedes proper analysis.\
Survival was good for B, C, and D.\
Pclass 1 stayed primarily in C and B.\
Pclass 2 and 3 stayed in F, D, E, and G.

In [8]:
def alone(row):
    if row.SibSp == 0 & row.Parch==0:
        return 'alone'
    else: 
        return 'not alone'

In [9]:
df['has_family'] = df.apply(alone, axis = 'columns')

In [10]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked deck has_family  
0      0         A/5 21171   7.2500   NaN        S  NaN  not alone  
1      0          PC 17599  71.2833   C85        C    C  not alone  
2      0  STON/O2. 3101282   7.9250   NaN        S  NaN      alone  
3      0            113803  53.1000  C123        S    C  not alone  
4      0            373450   8.0500   NaN        S  NaN      alone

In [11]:
df.groupby('has_family').count()

PassengerId  Survived  Pclass  Name  Sex  Age  SibSp  Parch  \
has_family                                                                
alone               608       608     608   608  608  471    608    608   
not alone           283       283     283   283  283  243    283    283   

            Ticket  Fare  Cabin  Embarked  deck  
has_family                                       
alone          608   608    126       606   126  
not alone      283   283     78       283    78

In [78]:
df.groupby(['survival_gender', 'Pclass']).Age.mean()

survival_gender  Pclass
female died      1         25.666667
                 2         36.000000
                 3         23.818182
female survived  1         34.939024
                 2         28.080882
                 3         19.329787
male died        1         44.581967
                 2         33.369048
                 3         27.255814
male survived    1         36.248000
                 2         16.022000
                 3         22.274211
Name: Age, dtype: float64

In [12]:
df.groupby(['has_family', 'Sex']).Pclass.value_counts()

has_family  Sex     Pclass
alone       female  3          81
                    1          49
                    2          44
            male    3         270
                    1          88
                    2          76
not alone   female  3          63
                    1          45
                    2          32
            male    3          77
                    1          34
                    2          32
Name: Pclass, dtype: int64

In [15]:
df.groupby(['Survived', 'has_family','Pclass']).Age.mean()

Survived  has_family  Pclass
0         alone       1         46.375000
                      2         33.833333
                      3         29.103175
          not alone   1         37.800000
                      2         32.870370
                      3         20.611111
1         alone       1         34.594203
                      2         29.333261
                      3         22.990333
          not alone   1         36.375849
                      2         21.635135
                      3         15.020000
Name: Age, dtype: float64

Being alone played a pretty large role on those who survived or not.

In [29]:
df.groupby(['Survived','Sex']).has_family.value_counts()

Survived  Sex     has_family
0         female  not alone      44
                  alone          37
          male    alone         361
                  not alone     107
1         female  alone         137
                  not alone      96
          male    alone          73
                  not alone      36
Name: has_family, dtype: int64

3. Did having a family member matter?\
Yes, 347 men travelling alone died while 121 with families died. The number of females who died traveling alone was 27 (the minimum) compared to the 134 who were traveling with someone and survived. 